### PR TITLE
Quantile Huber Loss to predict Specific Quantile Value

### PR DESCRIPTION
**Pull request by ashutosh1919**
We know that we have huber loss function added in keras tf-2 already which can perform both kind of behaviour MSE and MAE depending on the scale of data. In most of the prediction and analysis models, we often do not need just median or mean predicted value, but we also need the specific quantile value of prediction. This is the loss function I am adding which is derived from **huber_loss** here
[https://arxiv.org/pdf/1402.4624.pdf](https://arxiv.org/pdf/1402.4624.pdf) I name it **quantile_huber_loss** and the class which calls it is **QuantileHuber**. I have given great attention to the detail in implementing the function. function takes two extra arguements delta and quantile, where delta is common as the huber_loss and quantile is the float number between 0 and 1. Equation looks as below.
qhuber

### CODE REVIEW

**Comment by RomainBrault** [comment]
Related: #32871

**Comment by stefanmeili** [comment]
I'm not sure this loss function works as intended. At low or high tau (quantile) and and for larger kappa values, the intersection of the linear segments moves away from from the y axis. Using this loss will significantly bias the predicted quantile to a more extreme value than intended.

These plots explain the problem better.
bokeh_plot(1)

bokeh_plot(2)

**Comment by ashutosh1919** [comment]
> I'm not sure this loss function works as intended. At low or high tau (quantile) and and for larger kappa values, the intersection of the linear segments moves away from from the y axis. I suspect that using this loss will significantly bias the predicted quantile to a less extreme value than intended.
> 
> These plots explain the problem better.
> bokeh_plot(1)
> 
> bokeh_plot(2)

@stefanmeili, That is the sensitivity of this loss function. Training quality will largely depend upon tau and kappa values. But I have used this loss function to build forecasting engine. Is this loss function is used with accurate parameters, then it helps training of very dynamic data to forecast.

**Comment by stefanmeili** [comment]
Hi @ashutosh1919,
Thanks, I did look through the referenced paper and understand the motivation behind it. It just occurred to me that the quantile wouldn't be estimated correctly and thought it would be worth mentioning. I put together a quick numerical simulation via convolution to try to understand the magnitude of the discrepancy:
bokeh_plot(3)

The pinball loss correctly locates the 10% quantile at -1.2815. Quantile Huber loss estimates the 10% quantile at -0.9340 which translates to about the 17.5% quantile. This discrepancy grows for  quantiles closer to 0 or 1 and larger values of kappa.

I wonder if a horizontal offset can be applied to the error term (x in the paper) to correct for this? It probably won't be a pretty solution.
